File: sky/engine/core/painting/Color.dart
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

part of dart_ui;

Color _scaleAlpha(Color a, double factor) {
  return a.withAlpha((a.alpha * factor).round());
}

/// An immutable 32 bit color value in ARGB
class Color {
  /// Construct a color from the lower 32 bits of an int.
  ///
  /// Bits 24-31 are the alpha value.
  /// Bits 16-23 are the red value.
  /// Bits 8-15 are the green value.
  /// Bits 0-7 are the blue value.
  const Color(int value) : _value = (value & 0xFFFFFFFF);

  /// Construct a color from the lower 8 bits of four integers.
  const Color.fromARGB(int a, int r, int g, int b) :
    _value = ((((a & 0xff) << 24) |
                ((r & 0xff) << 16) |
                ((g & 0xff) << 8) |
                ((b & 0xff) << 0)) & 0xFFFFFFFF);

  /// A 32 bit value representing this color.
  ///
  /// Bits 24-31 are the alpha value.
  /// Bits 16-23 are the red value.
  /// Bits 8-15 are the green value.
  /// Bits 0-7 are the blue value.
  int get value => _value;
  final int _value;

  /// The alpha channel of this color in an 8 bit value.
  int get alpha => (0xff000000 & _value) >> 24;

  /// The alpha channel of this color as a double.
  double get opacity => alpha / 0xFF;

  /// The red channel of this color in an 8 bit value.
  int get red => (0x00ff0000 & _value) >> 16;

  /// The green channel of this color in an 8 bit value.
  int get green => (0x0000ff00 & _value) >> 8;

  /// The blue channel of this color in an 8 bit value.
  int get blue => (0x000000ff & _value) >> 0;

  /// Returns a new color that matches this color with the alpha channel
  /// replaced with a (which ranges from 0 to 255).
  Color withAlpha(int a) {
    return new Color.fromARGB(a, red, green, blue);
  }

  /// Returns a new color that matches this color with the alpha channel
  /// replaced with the given opacity (which ranges from 0.0 to 1.0).
  Color withOpacity(double opacity) {
    assert(opacity >= 0.0 && opacity <= 1.0);
    return withAlpha((255.0 * opacity).round());
  }

  /// Returns a new color that matches this color with the red channel replaced
  /// with r.
  Color withRed(int r) {
    return new Color.fromARGB(alpha, r, green, blue);
  }

  /// Returns a new color that matches this color with the green channel
  /// replaced with g.
  Color withGreen(int g) {
    return new Color.fromARGB(alpha, red, g, blue);
  }

  /// Returns a new color that matches this color with the blue channel replaced
  /// with b.
  Color withBlue(int b) {
    return new Color.fromARGB(alpha, red, green, b);
  }

  /// Linearly interpolate between two colors
  ///
  /// If either color is null, this function linearly interpolates from a
  /// transparent instance of the other color.
  static Color lerp(Color a, Color b, double t) {
    if (a == null && b == null)
      return null;
    if (a == null)
      return _scaleAlpha(b, t);
    if (b == null)
      return _scaleAlpha(a, 1.0 - t);
    return new Color.fromARGB(
      lerpDouble(a.alpha, b.alpha, t).toInt(),
      lerpDouble(a.red, b.red, t).toInt(),
      lerpDouble(a.green, b.green, t).toInt(),
      lerpDouble(a.blue, b.blue, t).toInt()
    );
  }

  @override
  bool operator ==(dynamic other) {
    if (other is! Color)
      return false;
    final Color typedOther = other;
    return value == typedOther.value;
  }

  @override
  int get hashCode => _value.hashCode;

  @override
  String toString() => "Color(0x${_value.toRadixString(16).padLeft(8, '0')})";
}

File: impeller/impeller/entity/Color.cc
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "Color.h"
#include <algorithm>
#include <cmath>
#include <sstream>

namespace rl {
namespace entity {

ColorHSB ColorHSB::FromRGB(Color rgb) {
  double R = rgb.red;
  double G = rgb.green;
  double B = rgb.blue;

  double v = 0.0;
  double x = 0.0;
  double f = 0.0;

  int64_t i = 0;

  x = fmin(R, G);
  x = fmin(x, B);

  v = fmax(R, G);
  v = fmax(v, B);

  if (v == x) {
    return ColorHSB(0.0, 0.0, v, rgb.alpha);
  }

  f = (R == x) ? G - B : ((G == x) ? B - R : R - G);
  i = (R == x) ? 3 : ((G == x) ? 5 : 1);

  return ColorHSB(((i - f / (v - x)) / 6.0), (v - x) / v, v, rgb.alpha);
}

Color ColorHSB::toRGBA() const {
  double h = hue * 6.0;
  double s = saturation;
  double v = brightness;

  double m = 0.0;
  double n = 0.0;
  double f = 0.0;

  int64_t i = 0;

  if (h == 0) {
    h = 0.01;
  }

  if (h == 0.0) {
    return Color(v, v, v, alpha);
  }

  i = static_cast<int64_t>(floor(h));

  f = h - i;

  if (!(i & 1)) {
    f = 1 - f;
  }

  m = v * (1 - s);
  n = v * (1 - s * f);

  switch (i) {
    case 6:
    case 0:
      return Color(v, n, m, alpha);
    case 1:
      return Color(n, v, m, alpha);
    case 2:
      return Color(m, v, n, alpha);
    case 3:
      return Color(m, n, v, alpha);
    case 4:
      return Color(n, m, v, alpha);
    case 5:
      return Color(v, m, n, alpha);
  }
  return Color(0, 0, 0, alpha);
}

std::string ColorHSB::toString() const {
  std::stringstream stream;
  stream << "{" << hue << ", " << saturation << ", " << brightness << ", "
         << alpha << "}";
  return stream.str();
}

Color::Color(const ColorHSB& hsbColor) : Color(hsbColor.toRGBA()) {}

std::string Color::toString() const {
  std::stringstream stream;
  stream << red << "," << green << "," << blue << "," << alpha;
  return stream.str();
}

void Color::fromString(const std::string& str) {
  std::stringstream stream(str);
  stream >> red;
  stream.ignore();
  stream >> green;
  stream.ignore();
  stream >> blue;
  stream.ignore();
  stream >> alpha;
}

}  // namespace entity
}  // namespace rl
